### PR TITLE
modernize tooling: 2024 edition

### PR DESCRIPTION
**Pull request by Xiphe**
Main motivation behind this was to split [`@oazapfts/runtime`](https://www.npmjs.com/package/@oazapfts/runtime) into a standalone package in order to allow for more efficient package handling.
While at it I moved to `"type": "module"` and made it so we provide `esm` and `cjs` builds.

- remove husky and lint-staged in favor of linting in ci in order to reduce tooling packages required
- switch package to type=module
- switch to vite in order to provide cjs and mjs entries
- switch to vitest less config, less packages, faster
- introduce packages/ folder #monorepostyle

BREAKING CHANGE:

package has been split and imports have been cleaned up

old                        | new
---------------------------|--------------------------
oazapfts/runtime           | @oazapfts/runtime
oazapfts/lib/runtime       | @oazapfts/runtime
oazapfts/runtime/query     | @oazapfts/runtime/query
oazapfts/lib/runtime/query | @oazapfts/runtime/query
$new -                   | @oazapfts/runtime/headers
$new -                   | @oazapfts/runtime/util
oazapfts/codegen           | oazapfts
oazapfts/lib/codegen       | oazapfts

non-generated code using oazapfts types and imports will most likely need to be migrated

generated clients should now have @oazapfts/runtime as dependency instead of oazapfts

`handle`, `ok`, `okify` and `optimistic` runtime helpers have been moved to `@oazapfts/runtime` along with other exports that mainly have been used internally. When you miss something from `oazapfts` package, try importing it from `@oazapfts/runtime`

ref https://github.com/oazapfts/oazapfts/pull/568
fix https://github.com/oazapfts/oazapfts/issues/401

### CODE REVIEW

**Comment by Xiphe** [comment]
Pre-released as [`oazapfts@6.0.0-beta.1`](https://github.com/oazapfts/oazapfts/releases/tag/oazapfts-v6.0.0-beta.1) and [`@oazapfts/runtime@1.0.0-beta.1`](https://github.com/oazapfts/oazapfts/releases/tag/%40oazapfts%2Fruntime-v1.0.0-beta.1)

**Comment by Xiphe** [comment]
Would appreciate if someone could give this a review. As this is quite a pack I'd also be available to pair it. 🤙

**Comment by Xiphe** [comment]
Was able to update one of my clients that I'm maintaining pretty seamlessly 🤙

File: packages/codegen/package.json [hunk 1]
```diff
@@ -0,0 +1,60 @@
+{
+  "name": "oazapfts",
+  "version": "0.0.0-development",
+  "description": "OpenApi TypeScript client generator",
+  "license": "MIT",
+  "repository": "oazapfts/oazapfts",
+  "type": "module",
+  "main": "dist/index.cjs",
+  "module": "dist/index.js",
```
Comment (benmccann) [inline packages/codegen/package.json:9]: `module` has no effect when `exports` is present
```suggestion
```

**Review by fimion — COMMENTED** [review]
I noticed a bug in the cjs build that i've pointed out here, but otherwise it seems good?

I have an experimental branch for our work repo where i've updated to the beta version, and everythng runs well.

File: packages/runtime/package.json [hunk 2]
```diff
@@ -0,0 +1,58 @@
+{
+  "name": "@oazapfts/runtime",
+  "version": "0.0.0-development",
+  "description": "Runtime for OpenApi TypeScript client generator",
+  "type": "module",
+  "main": "dist/index.cjs",
+  "module": "dist/index.js",
```
Comment (benmccann) [inline packages/runtime/package.json:7]: `module` will have no effect when `exports` is present
```suggestion
```

File: packages/codegen/src/generate.ts [hunk 3]
```diff
@@ -1,11 +1,12 @@
 import _ from "lodash";
-import ts, { factory } from "typescript";
-import path from "path";
+import ts from "typescript";
 import { OpenAPIV3, OpenAPIV3_1 } from "openapi-types";
 import * as cg from "./tscodegen";
 import generateServers, { defaultBaseUrl } from "./generateServers";
 import { Opts } from ".";
 
+const factory = ts.factory;
```
[the remark this answers is not in the document]
Comment (Xiphe) [reply packages/codegen/src/generate.ts:8]: oh, good catch! I'm gonna have a look 👍 

Thanks for reviewing! Highly appreciated 🫶

File: packages/runtime/package.json [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    },
+    "./util": {
+      "types": "./dist/util.d.ts",
+      "require": "./dist/util.cjs",
+      "import": "./dist/util.js"
+    },
+    "./headers": {
+      "types": "./dist/headers.d.ts",
+      "require": "./dist/headers.cjs",
+      "import": "./dist/headers.js"
```
Comment (benmccann) [inline packages/runtime/package.json:28]: I believe these are functionally equivalent, but I've found the error messages can be better with `default` than `import` when people do something wrong
```suggestion
      "require": "./dist/index.cjs",
      "default": "./dist/index.js"
    },
    "./query": {
      "types": "./dist/query.d.ts",
      "require": "./dist/query.cjs",
      "default": "./dist/query.js"
    },
    "./util": {
      "types": "./dist/util.d.ts",
      "require": "./dist/util.cjs",
      "default": "./dist/util.js"
    },
    "./headers": {
      "types": "./dist/headers.d.ts",
      "require": "./dist/headers.cjs",
      "default": "./dist/headers.js"
```

File: packages/runtime/package.json [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+    }
+  },
+  "license": "MIT",
+  "repository": "oazapfts/oazapfts",
```
Comment (benmccann) [inline packages/runtime/package.json:32]: it's nice to include a `directory` here
```suggestion
  "repository": {
    "type": "git",
    "url": "https://github.com/oazapfts/oazapfts",
    "directory": "packages/runtime"
  },
```

File: packages/codegen/package.json [hunk 6]
```diff
@@ -0,0 +1,60 @@
+{
+  "name": "oazapfts",
+  "version": "0.0.0-development",
+  "description": "OpenApi TypeScript client generator",
+  "license": "MIT",
+  "repository": "oazapfts/oazapfts",
```
Comment (benmccann) [inline packages/codegen/package.json:6]: it's nice to include a `directory` here for monorepos
```suggestion
  "repository": {
    "type": "git",
    "url": "https://github.com/oazapfts/oazapfts",
    "directory": "packages/codegen"
  },
```

**Review by benmccann — COMMENTED** [review]
lgtm! thanks again for all your work on this library

**Comment by Xiphe** [comment]
Awesome! Thank you both for the feedback and suggestions!

File: packages/codegen/src/generate.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (fimion) [reply packages/codegen/src/generate.ts:8]: beta.2 fixes this.

**Review by fimion — APPROVED** [review]
Approved!

threddy-threddyrex approved!

File: packages/codegen/src/generate.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (Xiphe) [reply packages/codegen/src/generate.ts:8]: Awesome!

**Pull request merged**